From a5394af741fc1746a9edab72160d0f41a05c30e9 Mon Sep 17 00:00:00 2001
From: Ranjani Vaidyanathan <ranjani.vaidyanathan@freescale.com>
Date: Thu, 31 Jul 2008 15:39:41 -0500
Subject: [PATCH] ENGR00086176 Change I2C bus driver to wait and retry to get access to the bus.

Also changed the parent of I2C to be ipg_perclk instead of ipg_clk.

Signed-off-by: Ranjani Vaidyanathan <ranjani.vaidyanathan@freescale.com>
---
 arch/arm/mach-mx37/clock.c   |    6 +++---
 drivers/i2c/busses/mxc_i2c.c |   10 +++++++++-
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-mx37/clock.c b/arch/arm/mach-mx37/clock.c
index 4649fa0..eeaad96 100644
--- a/arch/arm/mach-mx37/clock.c
+++ b/arch/arm/mach-mx37/clock.c
@@ -805,7 +805,7 @@ static struct clk i2c_clk[] = {
 	{
 	 .name = "i2c_clk",
 	 .id = 0,
-	 .parent = &ipg_clk,
+	 .parent = &ipg_perclk,
 	 .enable_reg = MXC_CCM_CCGR1,
 	 .enable_shift = MXC_CCM_CCGR1_CG14_OFFSET,
 	 .enable = _clk_enable,
@@ -814,7 +814,7 @@ static struct clk i2c_clk[] = {
 	{
 	 .name = "i2c_clk",
 	 .id = 1,
-	 .parent = &ipg_clk,
+	 .parent = &ipg_perclk,
 	 .enable_reg = MXC_CCM_CCGR1,
 	 .enable_shift = MXC_CCM_CCGR1_CG15_OFFSET,
 	 .enable = _clk_enable,
@@ -823,7 +823,7 @@ static struct clk i2c_clk[] = {
 	{
 	 .name = "i2c_clk",
 	 .id = 2,
-	 .parent = &ipg_clk,
+	 .parent = &ipg_perclk,
 	 .enable_reg = MXC_CCM_CCGR2,
 	 .enable_shift = MXC_CCM_CCGR2_CG0_OFFSET,
 	 .enable = _clk_enable,
diff --git a/drivers/i2c/busses/mxc_i2c.c b/drivers/i2c/busses/mxc_i2c.c
index 1ce269e..73ed2bb 100644
--- a/drivers/i2c/busses/mxc_i2c.c
+++ b/drivers/i2c/busses/mxc_i2c.c
@@ -430,11 +430,19 @@ static int mxc_i2c_xfer(struct i2c_adapter *adap, struct i2c_msg msgs[],
 	/*
 	 * Check bus state
 	 */
-	if (sr & MXC_I2SR_IBB) {
+
+	int retry = 5;
+	if (sr & MXC_I2SR_IBB && retry--) {
+		udelay(5);
+		sr = readw(dev->membase + MXC_I2SR);
+	}
+
+	if (sr & MXC_I2SR_IBB && retry < 0) {
 		mxc_i2c_module_dis(dev);
 		printk(KERN_DEBUG "Bus busy\n");
 		return -EREMOTEIO;
 	}
+
 	//gpio_i2c_active(dev->adap.id);
 	dev->transfer_done = false;
 	dev->tx_success = false;
-- 
1.5.4.4

